File: sdk/python/workswithagents/sla.py
"""
Agent SLA Framework — L7 Governance
Define and track Service Level Agreements for agent fleets.

Usage:
    from workswithagents import SLAMetrics
    
    sla = SLAMetrics("my-fleet", tier="production")
    
    # Report after each task
    sla.report("my-agent", "task-42", duration_seconds=187, success=True)
    
    # Check SLA status
    status = sla.status()  # {breaches: [], status: "ok"}
    
    # Generate monthly report
    report = sla.report_monthly("2026-05")
"""

import json
import time
import urllib.request
import urllib.error
from typing import Optional

DEFAULT_API = "https://workswithagents.dev"


class SLAMetrics:
    """Track and report SLA metrics for agent fleets."""
    
    TIERS = {
        "best_effort": {
            "uptime": 0.95,
            "accuracy": 0.80
        },
        "production": {
            "uptime": 0.995,
            "accuracy": 0.90,
            "latency_p95": 300,       # seconds
            "recovery": 0.95
        },
        "regulated": {
            "uptime": 0.999,
            "accuracy": 0.95,
            "latency_p99": 120,       # seconds
            "compliance": 1.0,        # all ATP-3 actions must pass
            "recovery": 0.99,
            "audit_retention_days": 2555  # 7 years
        }
    }
    
    def __init__(self, fleet_id: str, tier: str = "production", 
                 api_url: str = DEFAULT_API):
        self.fleet_id = fleet_id
        self.tier = tier
        self.api = api_url.rstrip("/")
        self.targets = self.TIERS.get(tier, self.TIERS["production"])
    
    def report(self, agent_id: str, action_id: str, *,
               duration_seconds: float, success: bool,
               guarantee_level: Optional[str] = None) -> dict:
        """Report a completed action for SLA tracking."""
        payload = {
            "fleet_id": self.fleet_id,
            "agent_id": agent_id,
            "action_id": action_id,
            "duration_seconds": duration_seconds,
            "success": success,
            "timestamp": int(time.time())
        }
        if guarantee_level:
            payload["guarantee_level"] = guarantee_level
        
        return self._request("POST", "/v1/sla/report", payload)
    
    def status(self) -> dict:
        """Get current SLA status — breaches and metrics."""
        return self._request("GET", f"/v1/sla/{self.fleet_id}/status")
    
    def report_monthly(self, period: str) -> dict:
        """Generate monthly SLA report (format: '2026-05')."""
        return self._request("GET", 
            f"/v1/sla/{self.fleet_id}/report?period={period}")
    
    def check_breach(self, metric: str, actual: float) -> Optional[str]:
        """Check if a metric breaches its target. Returns breach message or None."""
        target = self.targets.get(metric)
        if target is None:
            return None
        if actual < target:
            return f"{metric}: {actual:.3f} < target {target:.3f}"
        return None
    
    def _request(self, method: str, path: str, data: Optional[dict] = None) -> dict:
        url = f"{self.api}{path}"
        body = json.dumps(data).encode() if data else None
        req = urllib.request.Request(url, data=body, method=method)
        req.add_header("Content-Type", "application/json")
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            return {"error": e.code, "message": e.read().decode()}

File: sdk/typescript/src/sla.ts
/**
 * Agent SLA Framework — Layer 7 Governance (TypeScript)
 * Define and track Service Level Agreements for agent fleets.
 * Reference implementation. CC BY 4.0.
 */

const DEFAULT_API = "https://workswithagents.dev";

export type SLATier = "best_effort" | "production" | "regulated";

export interface SLATargets {
  uptime?: number;
  accuracy?: number;
  latency_p95?: number;
  latency_p99?: number;
  recovery?: number;
  compliance?: number;
  audit_retention_days?: number;
}

const TIERS: Record<SLATier, SLATargets> = {
  best_effort: { uptime: 0.95, accuracy: 0.80 },
  production: { uptime: 0.995, accuracy: 0.90, latency_p95: 300, recovery: 0.95 },
  regulated: { uptime: 0.999, accuracy: 0.95, latency_p99: 120, compliance: 1.0, recovery: 0.99, audit_retention_days: 2555 },
};

export class SLAMetrics {
  fleet_id: string;
  tier: SLATier;
  api: string;
  targets: SLATargets;

  constructor(fleet_id: string, tier: SLATier = "production", api_url: string = DEFAULT_API) {
    this.fleet_id = fleet_id;
    this.tier = tier;
    this.api = api_url.replace(/\/$/, "");
    this.targets = TIERS[tier] || TIERS.production;
  }

  async report(
    agent_id: string,
    action_id: string,
    duration_seconds: number,
    success: boolean,
    guarantee_level?: string
  ): Promise<Record<string, unknown>> {
    const payload: Record<string, unknown> = {
      fleet_id: this.fleet_id,
      agent_id,
      action_id,
      duration_seconds,
      success,
      timestamp: Math.floor(Date.now() / 1000),
    };
    if (guarantee_level) payload.guarantee_level = guarantee_level;
    return this.request("POST", "/v1/sla/report", payload);
  }

  async status(): Promise<Record<string, unknown>> {
    return this.request("GET", `/v1/sla/${this.fleet_id}/status`);
  }

  async reportMonthly(period: string): Promise<Record<string, unknown>> {
    return this.request("GET", `/v1/sla/${this.fleet_id}/report?period=${period}`);
  }

  checkBreach(metric: string, actual: number): string | null {
    const target = (this.targets as any)[metric];
    if (target == null) return null;
    if (actual < target) return `${metric}: ${actual.toFixed(3)} < target ${target.toFixed(3)}`;
    return null;
  }

  private async request(method: string, path: string, data?: Record<string, unknown>): Promise<any> {
    const url = `${this.api}${path}`;
    const opts: RequestInit = { method, headers: { "Content-Type": "application/json" } };
    if (data) opts.body = JSON.stringify(data);
    const resp = await fetch(url, opts);
    return resp.json();
  }
}
